### PR TITLE
AsyncCache

### LINKED ISSUE #15327: AsycCache

### PR DESCRIPTION
**Pull request by ausfeldt**
PR for visibility. Still need to handle errors.

Closes #15327

### CODE REVIEW

File: sdk/cosmos/azcosmos/async_cache.go [hunk 1]
```diff
@@ -0,0 +1,142 @@
...
+	fn       cacheValueTask
+	ch       <-chan interface{}
+}
+
+func newAsyncCache() *asyncCache {
+	return &asyncCache{}
+}
+
+func (ac *asyncCache) set(key interface{}, value interface{}) {
+	ac.values.Store(key, cacheValue{value: value})
```
Comment (ealsur) [inline sdk/cosmos/azcosmos/async_cache.go:34]: Does the sync.Map take care of concurrent sets for the same key?

File: sdk/cosmos/azcosmos/async_cache.go [hunk 2]
```diff
@@ -0,0 +1,142 @@
...
+}
+
+func (ac *asyncCache) setAsync(key interface{}, singleValueInit cacheValueTask, ctx context.Context) error {
+	ch := ac.execCacheValueTask(singleValueInit)
```
Comment (ealsur) [inline sdk/cosmos/azcosmos/async_cache.go:38]: How does the flow work on concurrent calls to setAsync? Wouldn't both create a channel and both call Store?

**Review by ealsur — APPROVED** [review]
Nothing blocking, we will probably find out more with usage, just trying to understand how certain code paths work

File: sdk/cosmos/azcosmos/async_cache.go [hunk 3]
```diff
@@ -0,0 +1,161 @@
...
+	_, err := ac.awaitCacheValue(key, ctx)
+
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (ac *asyncCache) getValue(key interface{}) (interface{}, bool) {
```
Comment (ealsur) [inline sdk/cosmos/azcosmos/async_cache.go:56]: nit: the sets are set and setAsync, for consistency, get and getAsync

File: sdk/cosmos/azcosmos/async_cache.go (hunk 1 above)
Comment (ausfeldt) [reply sdk/cosmos/azcosmos/async_cache.go:40]: sync.Map should take care of concurrent sets as it uses a RWMutex under the hood.

File: sdk/cosmos/azcosmos/async_cache.go (hunk 2 above)
Comment (ausfeldt) [reply sdk/cosmos/azcosmos/async_cache.go:44]: set is synchronous and changed the name to reflect this.  set is used to initialize the cache. getAsync is used to asynchronously get the value from the cache and update the value depending on the obsolete value passed.

**Pull request merged**